src/service/fixed-layer.js

```diff
@@ -686,116 +856,69 @@ export class FixedLayer {
+    this.layer_.id = 'i-amphtml-fixed-layer';
+    setImportantStyles(this.layer_, {
       position: 'absolute',
       top: 0,
       left: 0,
       height: 0,
       width: 0,
-      pointerEvents: 'none',
       overflow: 'hidden',
-
-      // Reset possible BODY styles.
-      animation: 'none',
-      background: 'none',
-      border: 'none',
-      borderImage: 'none',
-      boxSizing: 'border-box',
-      boxShadow: 'none',
-      display: 'block',
-      float: 'none',
-      margin: 0,
-      opacity: 1,
-      outline: 'none',
-      padding: 'none',
-      transform: 'none',
-      transition: 'none',
-      visibility: 'visible',
     });
-    doc.documentElement.appendChild(this.transferLayer_);
-    return this.transferLayer_;
+    doc.body.shadowRoot.appendChild(this.layer_);
   }
 
-  /**
-   * Find all `position:fixed` and `position:sticky` elements.
-   * @param {!Array<CSSRule>} rules
-   * @param {!Array<string>} foundSelectors
-   * @param {!Array<string>} stickySelectors
-   * @private
-   */
-  discoverSelectors_(rules, foundSelectors, stickySelectors) {
-    for (let i = 0; i < rules.length; i++) {
-      const rule = rules[i];
-      if (rule.type == /* CSSStyleRule */ 1) {
-        if (rule.selectorText != '*' && rule.style.position) {
-          if (rule.style.position == 'fixed') {
-            foundSelectors.push(rule.selectorText);
-          } else if (endsWith(rule.style.position, 'sticky')) {
-            stickySelectors.push(rule.selectorText);
-          }
-        }
-      } else if (rule.type == /* CSSMediaRule */ 4) {
-        this.discoverSelectors_(rule.cssRules, foundSelectors, stickySelectors);
-      } else if (rule.type == /* CSSSupportsRule */ 12) {
-        this.discoverSelectors_(rule.cssRules, foundSelectors, stickySelectors);
-      }
-    }
+  /** @override */
+  setVisible(visible) {
+    setStyle(this.layer_, 'visibility', visible ? 'visible' : 'hidden');
   }
-}
 
+  /** @override */
+  update() {
+    // Nothing to do.
+  }
 
-/**
- * @typedef {{
- *   id: string,
- *   selectors: !Array,
- *   element: !Element,
- *   position: string,
- *   placeholder: (?Element|undefined),
- *   fixedNow: boolean,
- *   stickyNow: boolean,
- *   top: (string|undefined),
- *   transform: (string|undefined),
- *   forceTransfer: (boolean|undefined),
- * }}
- */
-let ElementDef;
+  /** @override */
+  transferTo(fe) {
+    const {element} = fe;
+    const slotId = `i-amphtml-fixedid-${fe.id}`;
 
-/**
- * @typedef {{
- *   fixed: boolean,
- *   sticky: boolean,
- *   transferrable: boolean,
- *   top: string,
- *   zIndex: string,
- * }}
- */
-let ElementStateDef;
+    dev().fine(TAG, 'transfer to fixed:', fe.id, fe.element);
+    user().warn(TAG, 'In order to improve scrolling performance in Safari,' +
+        ' we now move the element to a fixed positioning layer:', fe.element);
+
+    // If never been transfered before: create and configure the slot.
+    if (!fe.slot) {
```

Comment:
Is a new slot necessary for every fixed element? Or could we just share a single slot? Chrome, at least, allows multiple elements in the same slot.

Reply:
OMG! This is amazing! Had no idea. Works very well.

Reply:
Done